lib/mongo/server/connection_pool/available_stack.rb

```diff
@@ -300,31 +288,83 @@ def close_stale_sockets!
           end
 
           mutex.synchronize do
-            num_checked_out = pool_size - queue_size
+            num_checked_out = pool_size - stack_size
             min_size_delta = [(min_size - num_checked_out), 0].max
 
             to_refresh.each do |connection|
-              if queue.include?(connection)
+              if connections.include?(connection)
                 connection.disconnect!
-                if queue.index(connection) < min_size_delta
+                if connections.index(connection) < min_size_delta
                   begin; connection.connect!; rescue; end
                 end
               end
             end
           end
         ensure
-          check_count_invariants
+          check_count_invariants(false)
+        end
+
+        def connection_removed
+          @pool_size -= 1
+          if @pool_size < 0
+            # This should never happen
+            log_warn("ConnectionPool::AvailableStack: unexpected push")
+            @pool_size = 0
+          end
         end
 
         private
 
-        def dequeue_connection
-          deadline = Time.now + wait_timeout
+        def close_connection!(connection, reason)
```

Comment:
A banged method name signifies a behavior difference from the non-banged method, or that something unusual is going on in the banged method. Simply closing objects is an insufficient reason to add bangs to method names.

Reply:
Changed